source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-custom-data-types.rst

```diff
@@ -0,0 +1,151 @@
+Creating Custom Data Types
+==========================
+
+Widgets allow us to control and visualize different types of data. This data can be numerical in the form to integers, double, longs, etc..., or less primitive forms of data, such as Java objects, or data involving multiple fields. In order to display these types of data using widgets, it is helpful to create a container class for them.
+It is not nessecary to create your own Data Class if the widget will handle single fieled data types such as doubles, arrays, or strings.
+
+Creating The Data Class
+-----------------------
+In this example, we will create a custom data type for a 2D Point and its x and y coordinates. In order to create a custom data type class,
+it must extend the abstract class `ComplexData <https://github.com/wpilibsuite/shuffleboard/blob/master/api/src/main/java/edu/wpi/first/shuffleboard/api/data/ComplexData.java>`_. Your
```

Comment:
Do all custom data types need to extend ComplexData?

Reply:
A custom data type should only be created if it involves multiple fields. A "SimpleData" class would simply mean that the data type only deals with one field in which case, we can use the [shuffleboard datatypes](https://github.com/wpilibsuite/shuffleboard/tree/76273d484c12a50356891b0ba7f7890b2ef25bc6/api/src/main/java/edu/wpi/first/shuffleboard/api/data/types) for whatever data type that one field is.